diff --git a/core/lib/Drupal/Core/Entity/ContentEntityType.php b/core/lib/Drupal/Core/Entity/ContentEntityType.php
index 3871405176..c27ee5e7c9 100644
--- a/core/lib/Drupal/Core/Entity/ContentEntityType.php
+++ b/core/lib/Drupal/Core/Entity/ContentEntityType.php
@@ -26,6 +26,14 @@ class ContentEntityType extends EntityType implements ContentEntityTypeInterface
    */
   protected $requiredRevisionMetadataKeys = [];
 
+  public function isStaticallyCacheable() {
+    return FALSE;
+  }
+
+  public function isRenderCacheable() {
+    return FALSE;
+  }
+
   /**
    * {@inheritdoc}
    */
