diff --git a/core/lib/Drupal/Core/Entity/ContentEntityType.php b/core/lib/Drupal/Core/Entity/ContentEntityType.php
index 3871405176..d14cd764a6 100644
--- a/core/lib/Drupal/Core/Entity/ContentEntityType.php
+++ b/core/lib/Drupal/Core/Entity/ContentEntityType.php
@@ -26,6 +26,18 @@ class ContentEntityType extends EntityType implements ContentEntityTypeInterface
    */
   protected $requiredRevisionMetadataKeys = [];
 
+  public function isStaticallyCacheable() {
+    return FALSE;
+  }
+
+  public function isRenderCacheable() {
+    return FALSE;
+  }
+
+  public function isPersistentlyCacheable() {
+    return FALSE;
+  }
+
   /**
    * {@inheritdoc}
    */
