diff --git a/lib/Drupal/Core/Cache/DatabaseBackend.php b/lib/Drupal/Core/Cache/DatabaseBackend.php
index 88f018f8cc..65cf3f7b87 100644
--- a/lib/Drupal/Core/Cache/DatabaseBackend.php
+++ b/lib/Drupal/Core/Cache/DatabaseBackend.php
@@ -110,7 +110,7 @@ class DatabaseBackend implements CacheBackendInterface {
     // ::select() is a much smaller proportion of the request.
     $result = [];
     try {
-      $result = $this->connection->query('SELECT cid, data, created, expire, serialized, tags, checksum FROM {' . $this->connection->escapeTable($this->bin) . '} WHERE cid IN ( :cids[] ) ORDER BY cid', [':cids[]' => array_keys($cid_mapping)]);
+      $result = $this->connection->query('SELECT SQL_NO_CACHE cid, data, created, expire, serialized, tags, checksum FROM {' . $this->connection->escapeTable($this->bin) . '} WHERE cid IN ( :cids[] ) ORDER BY cid', [':cids[]' => array_keys($cid_mapping)]);
     }
     catch (\Exception $e) {
       // Nothing to do.
