diff --git a/core/lib/Drupal/Core/Entity/EntityStorageBase.php b/core/lib/Drupal/Core/Entity/EntityStorageBase.php
index e5bf880aef..937aad4459 100644
--- a/core/lib/Drupal/Core/Entity/EntityStorageBase.php
+++ b/core/lib/Drupal/Core/Entity/EntityStorageBase.php
@@ -267,6 +267,16 @@ public function loadMultiple(array $ids = NULL) {
       }
       $entities = $passed_ids;
     }
+    foreach ($entities as $entity) {
+      if ($entity->label() == 'Ionizing radiation types') {
+        $backtrace = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 5);
+        array_shift($backtrace);
+        $backtrace = array_map(function ($value) {
+          return (!empty($value['file']) ? substr($value['file'], -50) : 'no file') . ':' . (!empty($value['line']) ? $value['line'] : 'no line');
+        }, $backtrace);
+        drupal_set_message(implode('|', $backtrace));
+      }
+    }
 
     return $entities;
   }
