diff --git a/src/Drupal/Commands/config/ConfigCommands.php b/src/Drupal/Commands/config/ConfigCommands.php
index ccc6755771..16117435a5 100644
--- a/src/Drupal/Commands/config/ConfigCommands.php
+++ b/src/Drupal/Commands/config/ConfigCommands.php
@@ -313,11 +313,18 @@ public function status($options = ['state' => 'Only in DB,Only in sync dir,Diffe
             ];
         }
 
-        if ($rows) {
-            return new RowsOfFields($rows);
-        } else {
+        if (!$rows) {
             $this->logger()->notice(dt('No differences between DB and sync directory.'));
+
+            // Suppress output if there are no differences and we are using the
+            // human readable "table" formatter so that we not uselessly output
+            // empty table headers.
+            if ($options['format'] === 'table') {
+                return null;
+            }
         }
+
+        return new RowsOfFields($rows);
     }
 
     /**
