diff --git a/entity_legal.install b/entity_legal.install
index 01cf134..980bc68 100644
--- a/entity_legal.install
+++ b/entity_legal.install
@@ -70,7 +70,7 @@ function entity_legal_update_8100() {
 /**
  * Move the published version field in the legal document version entity.
  */
-function entity_legal_update_8200(array &$sandbox) {
+function entity_legal_update_8101(array &$sandbox) {
   if (!isset($sandbox['published'])) {
     // Create the 'published' base field.
     $manger = \Drupal::entityDefinitionUpdateManager();
@@ -113,3 +113,10 @@ function entity_legal_update_8200(array &$sandbox) {
 
   $sandbox['#finished'] = (int) empty($sandbox['published']);
 }
+
+/**
+ * Set the database schema to 8200.
+ */
+function entity_legal_update_8200() {
+  // Nothing to do.
+}
