diff --git a/js/dropdown-widget.js b/js/dropdown-widget.js
index 695033a..eaee1b6 100644
--- a/js/dropdown-widget.js
+++ b/js/dropdown-widget.js
@@ -52,7 +52,7 @@
         .text(default_option_label);
       $dropdown.append($default_option);
 
-      $ul.prepend('<li class="default-option"><a href=".">' + default_option_label + '</a></li>');
+      $ul.prepend('<li class="default-option"><a href=".">' + Drupal.checkPlain(default_option_label) + '</a></li>');
 
       var has_active = false;
       $links.each(function () {
