diff --git a/src/OgDeleteOrphansBase.php b/src/OgDeleteOrphansBase.php
index 64adc409..aeb316c6 100644
--- a/src/OgDeleteOrphansBase.php
+++ b/src/OgDeleteOrphansBase.php
@@ -135,6 +135,18 @@ abstract class OgDeleteOrphansBase extends PluginBase implements OgDeleteOrphans
       return;
     }
 
+    if ($entity->getEntityTypeId() === 'og_membership' && $entity_id == 172) {
+      /** @var OgMembershipInterface $entity */
+      $group = $entity->getGroup();
+      $user = $entity->getOwner();
+      $roles = $entity->getRolesIds();
+      var_dump($group->label());
+      var_dump($user->label());
+      var_dump($roles);
+      ob_flush();
+    }
+
+
     // Only delete content that is fully orphaned, i.e. it is no longer
     // associated with any groups.
     $group_count = $this->membershipManager->getGroupCount($entity);
