diff --git a/src/OgDeleteOrphansBase.php b/src/OgDeleteOrphansBase.php
index 64adc409..29db16d2 100644
--- a/src/OgDeleteOrphansBase.php
+++ b/src/OgDeleteOrphansBase.php
@@ -129,6 +129,17 @@ abstract class OgDeleteOrphansBase extends PluginBase implements OgDeleteOrphans
   protected function deleteOrphan($entity_type, $entity_id) {
     $entity = $this->entityTypeManager->getStorage($entity_type)->load($entity_id);
 
+    if ($entity->getEntityTypeId() === 'og_membership' && $entity_id == 172) {
+      /** @var OgMembershipInterface $entity */
+      $group = $entity->getGroup();
+      $user = $entity->getOwner();
+      $roles = $entity->getRolesIds();
+      var_dump($group->label());
+      var_dump($user->label());
+      var_dump($roles);
+      ob_flush();
+    }
+
     // The entity might already be removed by other modules that implement
     // hook_entity_delete().
     if (!$entity) {
