diff --git a/og.module b/og.module
index f896775..eaa68c9 100755
--- a/og.module
+++ b/og.module
@@ -71,6 +71,16 @@ function og_entity_predelete(EntityInterface $entity) {
 }
 
 /**
+ * Implements hook_entity_delete().
+ */
+function og_entity_delete(Drupal\Core\Entity\EntityInterface $entity) {
+  // Clear relevant caches after a group or group content entity is deleted.
+  if (Og::isGroup($entity->getEntityTypeId(), $entity->bundle()) || Og::isGroupContent($entity->getEntityTypeId(), $entity->bundle())) {
+    Og::invalidateCache();
+  }
+}
+
+/**
  * Implements hook_entity_access().
  */
 function og_entity_access(EntityInterface $entity, $operation, AccountInterface $account) {
